File: routes/ecr.js
const express = require('express');
const { exec } = require('child_process');

const router = express.Router();

// ECR_PROFILE_MAP: maps AWS account ID → AWS SSO profile name (JSON)
// e.g. {"REDACTED_ACCOUNT":"PowerUserAccess-REDACTED_ACCOUNT"}
const ECR_PROFILE_MAP = (() => {
  try {
    return JSON.parse(process.env.ECR_PROFILE_MAP || '{}');
  } catch {
    return {};
  }
})();

/**
 * GET /api/ecr/tags?image=<full-ecr-image-url>
 * Parses ECR repo name + region from the image URL,
 * then calls `aws ecr describe-images` to return the 10 most recent tags.
 */
router.get('/ecr/tags', (req, res) => {
  const image = req.query.image;
  if (!image) {
    return res.status(400).json({ tags: [], error: 'Missing image query parameter' });
  }

  // Parse ECR image URL: <account>.dkr.ecr.<region>.amazonaws.com/<repo>:<tag>
  const match = image.match(/^(\d+)\.dkr\.ecr\.([^.]+)\.amazonaws\.com\/([^:]+)/);
  if (!match) {
    return res.status(400).json({ tags: [], error: 'Not a valid ECR image URL' });
  }

  const accountId = match[1];
  const region = match[2];
  const repository = match[3];

  // Look up profile for this account
  const profile = ECR_PROFILE_MAP[accountId];
  const env = profile ? { ...process.env, AWS_PROFILE: profile } : process.env;

  const cmd = `aws ecr describe-images --repository-name ${repository} --region ${region} --query 'sort_by(imageDetails,&imagePushedAt)[-10:]' --output json --no-cli-pager`;

  exec(cmd, { timeout: 60000, maxBuffer: 10 * 1024 * 1024, shell: '/bin/zsh', env }, (error, stdout, stderr) => {
    if (error) {
      console.error('ECR describe-images error:', stderr || error.message);
      return res.json({ tags: [], repository, error: stderr || error.message });
    }

    try {
      const imageDetails = JSON.parse(stdout);
      // Flatten all imageTags from the results, newest first
      const tags = imageDetails
        .reverse()
        .flatMap(detail => detail.imageTags || [])
        .filter(tag => tag); // remove empty

      res.json({ tags, repository });
    } catch (parseError) {
      res.json({ tags: [], repository, error: 'Failed to parse ECR response' });
    }
  });
});

module.exports = router;

File: cmd/server/routes/ecr.go
package routes

import (
	"encoding/json"
	"net/http"
	"os"
	"os/exec"
	"regexp"
	"strings"
)

var ecrURLPattern = regexp.MustCompile(`^(\d+)\.dkr\.ecr\.([^.]+)\.amazonaws\.com/([^:]+)`)

func registerECR(mux *http.ServeMux) {
	mux.HandleFunc("GET /api/ecr/tags", handleECRTags)
}

// GET /api/ecr/tags?image=<full-ecr-image-url>
// Parses ECR repo + region from image URL, calls aws ecr describe-images,
// returns 10 most recent tags. Triggered when user clicks a container image
// in pod detail view.
func handleECRTags(w http.ResponseWriter, r *http.Request) {
	image := r.URL.Query().Get("image")
	if image == "" {
		writeJSON(w, http.StatusBadRequest, map[string]any{"tags": []string{}, "error": "Missing image query parameter"})
		return
	}

	m := ecrURLPattern.FindStringSubmatch(image)
	if m == nil {
		writeJSON(w, http.StatusBadRequest, map[string]any{"tags": []string{}, "error": "Not a valid ECR image URL"})
		return
	}
	accountID, region, repository := m[1], m[2], m[3]

	// Account → SSO profile map from ECR_PROFILE_MAP env var.
	env := os.Environ()
	if profileMapJSON := os.Getenv("ECR_PROFILE_MAP"); profileMapJSON != "" {
		var profileMap map[string]string
		if json.Unmarshal([]byte(profileMapJSON), &profileMap) == nil {
			if profile, ok := profileMap[accountID]; ok {
				// Remove any existing AWS_PROFILE to avoid duplicates
				filtered := make([]string, 0, len(env))
				for _, e := range env {
					if !strings.HasPrefix(e, "AWS_PROFILE=") {
						filtered = append(filtered, e)
					}
				}
				env = append(filtered, "AWS_PROFILE="+profile)
			}
		}
	}

	cmd := exec.Command("aws",
		"ecr", "describe-images",
		"--repository-name", repository,
		"--region", region,
		"--query", "sort_by(imageDetails,&imagePushedAt)[-10:]",
		"--output", "json",
		"--no-cli-pager",
	)
	cmd.Env = env

	out, err := cmd.Output()
	if err != nil {
		var errMsg string
		if ee, ok := err.(*exec.ExitError); ok {
			errMsg = string(ee.Stderr)
		} else {
			errMsg = err.Error()
		}
		writeJSON(w, http.StatusOK, map[string]any{"tags": []string{}, "repository": repository, "error": errMsg})
		return
	}

	var imageDetails []struct {
		ImageTags []string `json:"imageTags"`
	}
	if err := json.Unmarshal(out, &imageDetails); err != nil {
		writeJSON(w, http.StatusOK, map[string]any{"tags": []string{}, "repository": repository, "error": "Failed to parse ECR response"})
		return
	}

	// Flatten tags newest-first, skip empty.
	tags := []string{}
	for i := len(imageDetails) - 1; i >= 0; i-- {
		for _, tag := range imageDetails[i].ImageTags {
			if tag != "" {
				tags = append(tags, tag)
			}
		}
	}

	writeJSON(w, http.StatusOK, map[string]any{"tags": tags, "repository": repository})
}
